tree/ntuple/src/RFieldUtils.cxx

```diff
@@ -118,37 +138,280 @@ std::vector<AnglePos> FindTemplateAngleBrackets(const std::string &typeName)
+      }
+   }
+}
+
+template <typename F>
+void NormalizeTemplateArguments(std::string &templatedTypeName, int maxTemplateArgs, F fnTypeNormalizer)
+{
+   const auto angleBrackets = FindTemplateAngleBrackets(templatedTypeName);
+   R__ASSERT(!angleBrackets.empty());
+
+   std::string normName;
+   std::string::size_type currentPos = 0;
+   for (std::size_t i = 0; i < angleBrackets.size(); i++) {
+      const auto [posOpen, posClose] = angleBrackets[i];
+      // Append the type prefix until the open angle bracket.
+      normName += templatedTypeName.substr(currentPos, posOpen + 1 - currentPos);
+
+      const auto argList = templatedTypeName.substr(posOpen + 1, posClose - posOpen - 1);
+      const auto templateArgs = ROOT::Internal::TokenizeTypeList(argList, maxTemplateArgs);
+      R__ASSERT(!templateArgs.empty());
+
+      const bool isUserClass = IsUserClass(templatedTypeName);
+      for (const auto &a : templateArgs) {
+         normName += GetNormalizedTemplateArg(a, isUserClass, fnTypeNormalizer) + ",";
+      }
+
+      normName[normName.size() - 1] = '>';
+      currentPos = posClose + 1;
+   }
+
+   // Append the rest of the type from the last closing angle bracket.
+   const auto lastClosePos = angleBrackets.back().second;
+   normName += templatedTypeName.substr(lastClosePos + 1);
+
+   templatedTypeName = normName;
+}
+
+// Given a type name normalized by ROOT Meta, return the type name normalized according to the RNTuple rules.
+std::string GetRenormalizedMetaTypeName(const std::string &metaNormalizedName)
+{
+   const std::string canonicalTypePrefix{ROOT::Internal::GetCanonicalTypePrefix(metaNormalizedName)};
+   // RNTuple resolves Double32_t for the normalized type name but keeps Double32_t for the type alias
+   // (also in template parameters)
+   if (canonicalTypePrefix == "Double32_t")
+      return "double";
+
+   if (canonicalTypePrefix.find('<') == std::string::npos) {
+      // If there are no templates, the function is done.
+      return canonicalTypePrefix;
+   }
+
+   std::string normName{canonicalTypePrefix};
+   NormalizeTemplateArguments(normName, 0 /* maxTemplateArgs */, GetRenormalizedMetaTypeName);
+
+   return normName;
+}
+
+// Given a demangled name ("normalized by the compiler"), return the type name normalized according to the
+// RNTuple rules.
+std::string GetRenormalizedDemangledTypeName(const std::string &demangledName, bool renormalizeStdString)
+{
+   std::string canonicalTypePrefix{demangledName};
+   RemoveCVQualifiers(canonicalTypePrefix);
+   RemoveLeadingKeyword(canonicalTypePrefix);
+   MapIntegerType(canonicalTypePrefix);
+
+   if (canonicalTypePrefix.find('<') == std::string::npos) {
+      // If there are no templates, the function is done.
+      return canonicalTypePrefix;
+   }
+   RemoveSpaceBefore(canonicalTypePrefix, '>');
+   RemoveSpaceAfter(canonicalTypePrefix, ',');
+   RemoveSpaceBefore(canonicalTypePrefix, ','); // MSVC fancies spaces before commas in the demangled name
```

Comment:
Should we try collapsing these three operations in one to avoid scanning the string three times?